Fix quantization in the effect engine (metronome effect)

daschuer opened the pull request:
This was originally a fix for #13611 but it turns out that the beat information provided into the effect engine itself was broken. I have fixed this at the root by refactoring GroupFeatureState and the "beat_next" and "beat_prev" COs.
Than I was able to finally fix the metronome, which works now also reliable during scratching.

Not sure if there is a real use case for it, but the metronome was perfectly to test that, and also helps verifying a beat grid by ear. It would be a real fun to play any sample and not only the metronome click. That could become our first "transport" effect. I will file a feature request. 

Open is to check/fix other effects. I want to have that reviewed and merged first before building on that. 
See commits for details.

daschuer committed:
Rename beat_fraction to beat_fraction_buffer_end to clarify that the value is related to the buffer end.

daschuer committed:
Skip copy if source is equal the destination the copy is unnecessary and __restrict__ assumption is violated.

daschuer committed:
Avoid unnecessary seconds round-trip in GroupFeatureState

daschuer committed:
Remove unused include

daschuer committed:
GroupFeatureState: std::optional

daschuer committed:
Add scratch_rate to GroupFeatureState to allow to quantize effects during scratching

daschuer committed:
Fix quantization with the metronome effect. It was one buffer off.

daschuer committed:
Introduce m_actual_speed with is what the scalers actually did.

This is required for determining the exact feat position in a buffer.

daschuer committed:
Don't snap "beat_next" and "beat_prev" to the nearest beat. This was a no longer required  workaround form the time, where the Mixxx transport was integer precise.

daschuer committed:
Fix the metronome effect during scratching using scratch_rate.

daschuer committed:
Don't use std::optional::value introduced in macOS 10.13

daschuer commented on src/effects/backends/builtin/metronomeeffect.cpp:
```diff
@@ -77,9 +77,8 @@ void MetronomeEffect::processChannel(
-        maxFrames = static_cast<SINT>(
-                engineParameters.sampleRate() * groupFeatures.beat_length_sec);
+    if (m_pSyncParameter->toBool() && groupFeatures.has_beat_length_frames) {
+        maxFrames = static_cast<SINT>(groupFeatures.beat_length_frames);
```
I have replaced it by double precision in a later commit.

daschuer committed:
default GroupFeatureState ctor

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

daschuer committed:
Improve readability in meronomeeffect.cpp

daschuer committed:
Fix use of GroupFeatureState in LoudnessContourEffect

daschuer commented:
Done.

Swiftb0y commented on src/effects/backends/builtin/metronomeeffect.cpp:
the changes here are really hard to follow because they mix offset and timing calculations with buffer indexing (mixing levels of abstraction) and thus I'm neither confident about the reliability nor safety of the code...

Do you think its feasible to reorganize it a little so its easier to understand?

daschuer replied:
I have tried to fix that by calculate everything in frames. Maybe I was not too successful or comments are missing. 
Can you give some hints what can be improved?

Swiftb0y replied:
I will rework it a little and open a PR to your branch ASAP.

Swiftb0y replied:
Please take a look https://github.com/daschuer/mixxx/pull/97

Swiftb0y committed:
refactor: move metronome sampledata into separate translation unit

Swiftb0y committed:
refactor: use higher-level `std::span` based logic

daschuer commented:
I have added the first commit of your branch from https://github.com/daschuer/mixxx/pull/97
Can we merge this first, before doing the std::span refactoring? 
I am not convinced that this reduces complexity, because it brings back the sample count where I am very happy to with the frame based counters.

Swiftb0y commented:
> Can we merge this first, before doing the std::span refactoring?

Sure, but unfortunately I encountered a couple bugs during testing:

1. Sync on the Metronome is off: expected it should continue clicking based on the fixed rate of the knob. Currently Observed: it only does not click at all, only when enabling the effect, and with its transient chopped of.
2. Sync on the Metronome is on but track has no beatgrid. Expected: not quite sure actually whether it should stay silent or just play as if sync was turned off. Observed: it just clicks once when enabling, though this time its the full sample.

Both of these issues are regressions.

daschuer committed:
Fix metronome without a sync enabled

daschuer commented:
The described issues are fixed now. It was caused by a faulty fix to avoid SINT overflow.

Swiftb0y commented:
I don't understand the fix... There also remains the issue where the first click is with sync disabled is muffled.

daschuer committed:
Avoid to fade in the metronome effect, because this would mangle the first click.

daschuer commented:
> There also remains the issue where the first click is with sync disabled is muffled.

This is a "feature", the fade in of the effect. But we can safely disable this here. 

I have removed a redundant check and added comments to the last summit.

Swiftb0y commented on src/effects/backends/builtin/metronomeeffect.cpp:
```diff
@@ -57,71 +58,87 @@ void MetronomeEffect::processChannel(
+                // We have a new beat before the current buffer ends
+                nextClickStart = bufferEnd - beatToBufferEnd;
+            }
+        } else {
+            nextClickStart = *groupFeatures.beat_length_frames;
```
Can this actually happen?

daschuer replied:
No, I have added it only for completeness. We may consider to use both in a optional struct to get around it.

Swiftb0y replied:
do you mean `beat_length_frames` and `beat_fraction_buffer_end` or `beat_fraction_buffer_end` and `scratch_rate`? because according to `BpmControl::collectFeatures` if `rate_ratio` is 0, we can get a `beat_fraction_buffer_end` but not a `beat_length_frames`

daschuer replied:
We always have it all in reality, because we do not allow a rate of 0 but it is possible via writing the CO. 
We should probably treat this case in the same way we treat scratch = 0

Swiftb0y replied:
yes, I agree. I implemented that in a follow up. will push shortly.

daschuer replied:
I am already on it.

Swiftb0y replied:
sure that works, but it doesn't eliminate the original case I asked about (which is when `beat_length_frames.has_value()` but `!beat_fraction_buffer_end.has_value()`).

daschuer replied:
This can not happen. But looking at the API it, can. `beat_fraction_buffer_end.has_value() == false` means the track has no value. In this case `beat_length_frames.has_value()` is always false.

Swiftb0y replied:
right, so that should be reflected in `GroupFeatureState`, do you agree?

daschuer replied:
I am not sure, I because we may consider a sound source that has bpm info but no phase info.

daschuer committed:
Combine frames and scratch_rate into one optional because they are always valid.

daschuer commented:
Done, please have a look.

Swiftb0y commented:
fyi, I got https://github.com/daschuer/mixxx/pull/97 into a state I'm happy with. Please take a look.

daschuer commented:
I have adopted the full span based approach. But I have not adopted the rest, because of the std::size_t / std::ptrdiff_t / SINT clash with gsl/span. I also must admit that I can not fully understand how your code works.

Swiftb0y commented:
I think the primary advantage of my span approach was not the fact that it uses span (it just made it easier for myself to do the cleanup I wanted to do). The real improvement was separating the messy sync code from the rest. But I'm okay with the current state as a temporary stopgap.

Swiftb0y reviewed (approved):
This LGTM for now. I will retarget the follow up at `main`.
